feat(integrations/calendly): implement integration

JustinBordage opened the pull request.

JustinBordage committed:
feat: setup initial integration scaffold

JustinBordage committed:
fix: Run pnpm install to add integration to lockfile

JustinBordage committed:
chore: remove empty dev dependancies causing CI fails

JustinBordage committed:
chore: add integration description to address failing CI test

JustinBordage committed:
style: add newline at end of package.json to address CI test

JustinBordage committed:
feat: add configuration schema for integration

JustinBordage committed:
chore: add feature flag to prevent deployment to production

JustinBordage committed:
feat: add placeholder implementation for the scheduleEvent action

JustinBordage committed:
feat: Implement Calendly api request handlers

JustinBordage committed:
feat: add initial implementation for webhook registration & deregistration

JustinBordage committed:
fix: incorrect allowed property value in webhookDetailsSchema

JustinBordage committed:
refactor: calendly api functions & schemas

JustinBordage committed:
chore: update type operator in imports to match linting rules

JustinBordage committed:
style: rename "getCurrentUserResponseSchema" to "getCurrentUserRespSchema" to align with other schema naming

JustinBordage committed:
refactor: setup logic

JustinBordage committed:
docs: add links to Calendly api docs for each schema

JustinBordage committed:
style: fix import order

JustinBordage committed:
feat: implement scheduleEvent action

JustinBordage committed:
refactor: move calendlyUri schema & type to common.ts

JustinBordage committed:
feat: add implementation of webhook event schemas

JustinBordage committed:
fix: CalendlyUri type import

JustinBordage committed:
Merge branch 'master' into jbord_calendly-implement-integration

JustinBordage committed:
feat: implement webhook events

JustinBordage committed:
chore: update code owner for calendly integration

JustinBordage committed:
Merge branch 'master' into jbord_calendly-implement-integration

JustinBordage committed:
fix: webhook event schemas

JustinBordage committed:
feat: add conversationId parsing to event

JustinBordage committed:
Merge branch 'jbord_calendly-implement-integration' of https://github.com/botpress/botpress into jbord_calendly-implement-integration

JustinBordage committed:
fix: add description & title for event schema properties

JustinBordage committed:
chore: add examples of location types in webhook event output data

JustinBordage committed:
chore: write hub.md

JustinBordage committed:
fix: hub.md formatting

JustinBordage committed:
fix: config instruction that should be indented

franklevasseur reviewed (commented):
LGTM so far;

Here's a first round of comments. I'll do another review once adressed

JustinBordage committed:
chore: move calendly api schemas & types to src folder

JustinBordage committed:
fix: remove type branding from definitions

JustinBordage committed:
style: make some comments in setup.ts single lined

JustinBordage committed:
refactor: change filenames and directories for event schemas

JustinBordage committed:
chore: remove unused types and exports

franklevasseur commented on integrations/calendly/definitions/actions.ts:
```diff
@@ -0,0 +1,18 @@
+import { nonBlankString } from './common'
+
+export const scheduleEventInputSchema = z.object({
+  conversationId: nonBlankString
```
**nitpick:** you can use a
```ts
.placeholder('{{ event.conversationId }}')
```

franklevasseur commented on integrations/calendly/definitions/events/index.ts:
```diff
@@ -0,0 +1,142 @@
+
+export const inviteeEventOutputSchema = z.object({
+  eventName: nonBlankString.describe('The name of the scheduled event').title('Scheduled Event Name'),
+  startTime: z.date().describe('The start time of the scheduled event').title('Start Time'),
```
**issue:** don't use `z.date()` here; use `z.string().datetime()` as its behavior is easier to predict

franklevasseur commented on integrations/calendly/definitions/events/index.ts:
```diff
@@ -0,0 +1,142 @@
+  scheduled_event: scheduledEventSchema,
+})
+
+export const inviteeEventSchema = z.object({
```
**chore:** This schema is not used in the integration's definition. Move this in the `src/` folder. The same goes for all other schemas, the `definitions/` is for schemas used in the `integration.definition.ts` file.


**note:** if the `integration.definiton.ts` file is small enough, you don't even need to have a `definitions/` folder

franklevasseur commented on integrations/calendly/src/calendly-api.ts:
```diff
@@ -0,0 +1,104 @@
+// ------ Status Codes ------
+const NO_CONTENT = 204 as const
+
+export const getCurrentUser = async (axiosClient: CalendlyClient): Promise<CalendlyTypes.GetCurrentUserResp> => {
```
**question:** why not a class ? This way you can group all your calendly API specific logic in a single `calendly.ts` or `calendly-api.ts` file

```ts
export class CalendlyClient {
  private _axiosClient: AxiosInstance

  public constructor(authToken: string, baseURL: string = BASE_URL) {
    this._axiosClient = axios.create({
      baseURL,
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${authToken}`,
      },
    })
  }

  public getCurrentUser = async (): Promise<CalendlyTypes.GetCurrentUserResp> => {
    const resp = await this._axiosClient.get<object>('/users/me')
    return CalendlyTypes.getCurrentUserRespSchema.parse(resp.data)
  }
  
  // other methods
}
```

franklevasseur reviewed (commented):
another round;

`definitions/` is for definitions only

JustinBordage committed:
refactor: move example to placeholder field

JustinBordage committed:
chore: move schemas not used in 'integration.definition.ts' out of the 'definitions' directory

JustinBordage committed:
refactor: move & rename 'events/index.ts' to 'events.ts'

This is because the file is small enough that it doesn't need a dedicated directory anymore

JustinBordage committed:
fix: 'inviteeEventOutputSchema' to use `z.string().datetime()` instead of `z.date()`

JustinBordage committed:
fix: date serialization for invitee event output

JustinBordage committed:
fix: remove routing_form_submission event from webhook subscriptions

- This is because the event requires the user to be the admin for the organization in order to subscribe to "organization" level events

JustinBordage committed:
feat: add scheduledEventUri to the inviteeEvent output

JustinBordage committed:
refactor: Calendly api client into a class

franklevasseur reviewed (dismissed):
1 comment; the rest lgtm

JustinBordage committed:
refactor: made api response schemas less constrained

JustinBordage committed:
refactor: simplify nullable property schema

JustinBordage committed:
feat: add passthroughs to calendly api response schemas

JustinBordage committed:
refactor: made calendly webhook data schemas less constrained

JustinBordage committed:
feat: add passthroughs to calendly webhook data payload schemas

JustinBordage committed:
refactor: made integration definition schemas less constrained

JustinBordage committed:
feat: implement improved error logging for api schema validation

JustinBordage committed:
Merge branch 'master' into jbord_calendly-implement-integration
